Sources/Commands/SwiftTool.swift

```diff
@@ -498,34 +498,38 @@ public class SwiftTool {
     func getAuthorizationProvider() throws -> AuthorizationProvider? {
         var providers = [AuthorizationProvider]()
         // netrc file has higher specificity than keychain so use it first
-        if let workspaceNetrc = try self.getNetrcConfig()?.get() {
-            providers.append(workspaceNetrc)
+        if let netrcConfigFile = try self.getNetrcConfigFile() {
+            providers.append(try NetrcAuthorizationProvider(path: netrcConfigFile, fileSystem: localFileSystem))
         }
         
         // TODO: add --no-keychain option to allow opt-out
 //#if canImport(Security)
 //        providers.append(KeychainAuthorizationProvider(observabilityScope: self.observabilityScope))
 //#endif
         
-        return providers.isEmpty ? nil : CompositeAuthorizationProvider(providers, observabilityScope: self.observabilityScope)
+        return providers.isEmpty ? .none : CompositeAuthorizationProvider(providers, observabilityScope: self.observabilityScope)
     }
 
-    func getNetrcConfig() -> Workspace.Configuration.Netrc? {
-        guard options.netrc else { return nil }
+    func getNetrcConfigFile() throws -> AbsolutePath? {
+        guard options.netrc else {
+            return .none
+        }
 
         if let configuredPath = options.netrcFilePath {
             guard localFileSystem.exists(configuredPath) else {
-                self.observabilityScope.emit(error: "Did not find .netrc file at \(configuredPath).")
-                return nil
+                throw StringError("Did not find .netrc file at \(configuredPath).")
             }
+            return configuredPath
+        }
 
-            return .init(path: configuredPath, fileSystem: localFileSystem)
-        } else {
-            let defaultPath = localFileSystem.homeDirectory.appending(component: ".netrc")
-            guard localFileSystem.exists(defaultPath) else { return nil }
-
-            return .init(path: defaultPath, fileSystem: localFileSystem)
+        // TODO: replace multiroot-data-file with explicit overrides
+        let localPath = try (options.multirootPackageDataFile ?? self.getPackageRoot()).appending(component: ".netrc")
```

Comment:
@mattt not sure what location you had in mind for the local case. for this PR I assumed the package's root directory, but we can consider other locations like `<root>/.swiftpm/configuration/.netrc`

Reply:
That's perfect. I think `<root>/.netrc` is the right place to look for a local override.

As for `.swiftpm/configuration`, I think that makes more sense for configuration that's specific to Swift Package Manager. Since `.netrc` files are a de facto standard, I wouldn't expect to see them there.